### ISSUE TITLE
Windows CI failures in deprecation logging feature

### ISSUE THREAD

**Issue by jsvd** [elastic/logstash#11307]
Example failure here: https://logstash-ci.elastic.co/job/elastic+logstash+7.x+multijob-windows-compatibility/67/console

Text:

```
11:41:44 org.logstash.log.DefaultDeprecationLoggerTest > testDeprecationLoggerWriteOut_nested FAILED
11:41:44     java.nio.file.AccessDeniedException: A:\logstash-core\build\logs\logstash-deprecation.log
11:41:44         at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:83)
11:41:44         at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97)
11:41:44         at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102)
11:41:44         at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:269)
11:41:44         at sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:108)
11:41:44         at java.nio.file.Files.deleteIfExists(Files.java:1165)
11:41:44         at org.logstash.log.LogTestUtils.deleteLogFile(LogTestUtils.java:32)
11:41:44         at org.logstash.log.DefaultDeprecationLoggerTest.setUp(DefaultDeprecationLoggerTest.java:38)
```

```
11:41:44 org.logstash.log.DefaultDeprecationLoggerTest > testDeprecationLoggerWriteOut_nested STANDARD_OUT
11:41:44     2019-11-14 11:41:44,895 Test worker ERROR Unable to create file build/logs/logstash-deprecation.log java.io.IOException: Access is denied
11:41:44     	at java.io.WinNTFileSystem.createFileExclusively(Native Method)
11:41:44     	at java.io.File.createNewFile(File.java:1012)
11:41:44     	at org.apache.logging.log4j.core.appender.rolling.RollingFileManager$RollingFileManagerFactory.createManager(RollingFileManager.java:651)
11:41:44     	at org.apache.logging.log4j.core.appender.rolling.RollingFileManager$RollingFileManagerFactory.createManager(RollingFileManager.java:631)
11:41:44     	at org.apache.logging.log4j.core.appender.AbstractManager.getManager(AbstractManager.java:113)
```

**Comment by andsel** [comment]
Trying to reproduce the (with standard build) on a win machine it doesn't happen, seems a flaky test that's not so easy to make more robust. Looking at the first exception on the Jenkins log:
```
11:41:44 org.logstash.log.DefaultDeprecationLoggerTest > testDeprecationLoggerWriteOut_root STARTED
11:41:44 
11:41:44 org.logstash.log.DefaultDeprecationLoggerTest > testDeprecationLoggerWriteOut_root PASSED
11:41:44 
11:41:44 org.logstash.log.DefaultDeprecationLoggerTest > testDeprecationLoggerWriteOut_nested STARTED
11:41:44 
11:41:44 org.logstash.log.DefaultDeprecationLoggerTest > testDeprecationLoggerWriteOut_nested STANDARD_OUT
11:41:44     2019-11-14 11:41:44,895 Test worker ERROR Unable to create file build/logs/logstash-deprecation.log java.io.IOException: Access is denied
11:41:44     	at java.io.WinNTFileSystem.createFileExclusively(Native Method)
11:41:44     	at java.io.File.createNewFile(File.java:1012)
11:41:44     	at org.apache.logging.log4j.core.appender.rolling.RollingFileManager$RollingFileManagerFactory.createManager(RollingFileManager.java:651)
11:41:44     	at org.apache.logging.log4j.core.appender.rolling.RollingFileManager$RollingFileManagerFactory.createManager(RollingFileManager.java:631)
11:41:44     	at org.apache.logging.log4j.core.appender.AbstractManager.getManager(AbstractManager.java:113)
11:41:44     	at org.apache.logging.log4j.core.appender.OutputStreamManager.getManager(OutputStreamManager.java:114)
11:41:44     	at org.apache.logging.log4j.core.appender.rolling.RollingFileManager.getFileManager(RollingFileManager.java:205)
11:41:44     	at org.apache.logging.log4j.core.appender.RollingFileAppender$Builder.build(RollingFileAppender.java:146)
11:41:44     	at org.apache.logging.log4j.core.appender.RollingFileAppender$Builder.build(RollingFileAppender.java:62)
11:41:44     	at org.apache.logging.log4j.core.config.plugins.util.PluginBuilder.build(PluginBuilder.java:122)
11:41:44     	at org.apache.logging.log4j.core.config.AbstractConfiguration.createPluginObject(AbstractConfiguration.java:1002)
11:41:44     	at org.apache.logging.log4j.core.config.AbstractConfiguration.createConfiguration(AbstractConfiguration.java:942)
11:41:44     	at org.apache.logging.log4j.core.config.AbstractConfiguration.createConfiguration(AbstractConfiguration.java:934)
11:41:44     	at org.apache.logging.log4j.core.config.AbstractConfiguration.doConfigure(AbstractConfiguration.java:552)
11:41:44     	at org.apache.logging.log4j.core.config.AbstractConfiguration.initialize(AbstractConfiguration.java:241)
11:41:44     	at org.logstash.log.LogstashConfigurationFactory.getConfiguration(LogstashConfigurationFactory.java:44)
11:41:44     	at org.logstash.log.LogstashConfigurationFactory.getConfiguration(LogstashConfigurationFactory.java:16)
11:41:44     	at org.apache.logging.log4j.core.config.ConfigurationFactory$Factory.getConfiguration(ConfigurationFactory.java:458)
11:41:44     	at org.apache.logging.log4j.core.config.ConfigurationFactory$Factory.getConfiguration(ConfigurationFactory.java:385)
11:41:44     	at org.apache.logging.log4j.core.config.ConfigurationFactory.getConfiguration(ConfigurationFactory.java:293)
11:41:44     	at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:647)
11:41:44     	at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:668)
11:41:44     	at org.apache.logging.log4j.core.LoggerContext.start(LoggerContext.java:253)
11:41:44     	at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:243)
11:41:44     	at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:45)
11:41:44     	at org.apache.logging.log4j.LogManager.getContext(LogManager.java:174)
11:41:44     	at org.apache.logging.log4j.core.LoggerContext.getContext(LoggerContext.java:214)
11:41:44     	at org.logstash.log.LogTestUtils.reloadLogConfiguration(LogTestUtils.java:25)
11:41:44     	at org.logstash.log.DefaultDeprecationLoggerTest.tearDown(DefaultDeprecationLoggerTest.java:43)
```

 The problem is that the first test run, execute successfully and then during the teardown deletes the log file (if exists) and `reloadLogConfiguration` which itself invoke a stop with a timeout; from Javadoc if the timeouts expires before log4j2 terminates his internal state, than _should_ return false *but* looking at the implementation https://github.com/apache/logging-log4j2/blob/log4j-2.12.1/log4j-core/src/main/java/org/apache/logging/log4j/core/LoggerContext.java#L352-L396 it always return true. So the suspect is that there is something asynchronous on the file  (or file descriptor) that is still running when the second test teardown run because the file still exists while the realod of log4j2 context wants to create a new copy

**Comment by andsel** [comment]
closed by #11311

**Comment by jsvd** [comment]
@andsel Either the same or a similar problem is still occurring: https://logstash-ci.elastic.co/job/elastic+logstash+master+multijob-windows-compatibility/95/console

```
11:18:38 org.logstash.log.DefaultDeprecationLoggerTest > testDeprecationLoggerWriteOut_root STARTED
11:18:38 
11:18:38 org.logstash.log.DefaultDeprecationLoggerTest > testDeprecationLoggerWriteOut_root FAILED
11:18:38     java.nio.file.FileSystemException: A:\logstash-core\build\logs\logstash-deprecation.log: The process cannot access the file because it is being used by another process.
11:18:38         at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:86)
11:18:38         at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97)
11:18:38         at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102)
11:18:38         at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:269)
11:18:38         at sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:108)
11:18:38         at java.nio.file.Files.deleteIfExists(Files.java:1165)
11:18:38         at org.logstash.log.LogTestUtils.deleteLogFile(LogTestUtils.java:32)
11:18:38         at org.logstash.log.DefaultDeprecationLoggerTest.tearDown(DefaultDeprecationLoggerTest.java:43)
11:18:38 
11:18:38 org.logstash.log.DefaultDeprecationLoggerTest > testDeprecationLoggerWriteOut_nested STARTED
11:18:38 
11:18:38 org.logstash.log.DefaultDeprecationLoggerTest > testDeprecationLoggerWriteOut_nested FAILED
11:18:38     java.nio.file.FileSystemException: A:\logstash-core\build\logs\logstash-deprecation.log: The process cannot access the file because it is being used by another process.
11:18:38         at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:86)
11:18:38         at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97)
11:18:38         at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102)
11:18:38         at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:269)
11:18:38         at sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:108)
11:18:38         at java.nio.file.Files.deleteIfExists(Files.java:1165)
11:18:38         at org.logstash.log.LogTestUtils.deleteLogFile(LogTestUtils.java:32)
11:18:38         at org.logstash.log.DefaultDeprecationLoggerTest.setUp(DefaultDeprecationLoggerTest.java:38)
11:18:38 
11:18:38     java.nio.file.FileSystemException: A:\logstash-core\build\logs\logstash-deprecation.log: The process cannot access the file because it is being used by another process.
11:18:38         at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:86)
11:18:38         at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97)
11:18:38         at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102)
11:18:38         at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:269)
11:18:38         at sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:108)
11:18:38         at java.nio.file.Files.deleteIfExists(Files.java:1165)
11:18:38         at org.logstash.log.LogTestUtils.deleteLogFile(LogTestUtils.java:32)
11:18:38         at org.logstash.log.DefaultDeprecationLoggerTest.tearDown(DefaultDeprecationLoggerTest.java:43)
11:18:38 
```

**Comment by andsel** [comment]
I think is almost the same, before we:
- reconfigure the log (so get access to the log file)
- delete the log file

while now
- delete the log file
- reconfigure the log

I think the ratio is the same, when we exit a test (doing tear down duties) the resources we try to free are already brought by other process. This seems strange  because:
- the process should be the same, the JVM that's running the test
- the test should be serial (or during the build we parallelize the execution?)

**Comment by jsvd** [comment]
for now can you try a little loop that tries for 10 times:
1. reconfigure log
2. if FileSystemException
   1. log the error
   2. sleep 10
   3. increase retry count
   4. retry unless retry count == 10

**Comment by andsel** [comment]
Closed definetively by #11399
